### PR TITLE
Support Futures via widgets

### PR DESCRIPTION
**Pull request by AndrewFerr**
Signed-off-by: Andrew Ferrazzutti <andrewf@element.io>

Dependencies:
- https://github.com/matrix-org/matrix-js-sdk/pull/4294
- https://github.com/matrix-org/matrix-widget-api/pull/92

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [x] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

File: src/embedded.ts [hunk 1]
```diff
@@ -280,7 +299,80 @@ export class RoomWidgetClient extends MatrixClient {
...
+     * @experimental This currently relies on an unstable MSC (MSC4140).
+     */
+    // eslint-disable-next-line
+    public async _unstable_sendStateFuture<K extends keyof StateEvents, F extends SendFutureRequestOpts>(
+        roomId: string,
+        futureOpts: F,
+        eventType: K,
+        content: StateEvents[K],
+        stateKey: string,
+    ): Promise<SendFutureResponse<F>>;
```
Comment (t3chguy) [inline src/embedded.ts:317]: This overload seems to not add anything?

Comment (AndrewFerr) [reply src/embedded.ts:317]: This type signature is needed to properly override the method from the parent class, and the one following this is needed to avoid some typechecking errors, as type narrowing of return types is finicky when conditional parameter types are involved.

But I can avoid using an overload if I simply cast the return value, so I'll go with that.

**Comment by AndrewFerr** [comment]
Putting this back into draft until the widget API is up-to-speed with the recent revisions to MSC4140.

File: src/embedded.ts [hunk 2]
```diff
@@ -280,7 +290,48 @@ export class RoomWidgetClient extends MatrixClient {
...
+        if (!(await this.doesServerSupportUnstableFeature(UNSTABLE_MSC4140_DELAYED_EVENTS))) {
+            throw Error("Server does not support the delayed events API");
+        }
+
+        const response = await this.widgetApi.sendStateEvent(
+            eventType,
+            stateKey,
+            content,
+            roomId,
+            "delay" in delayOpts ? delayOpts.delay : undefined,
```
Comment (toger5) [inline src/embedded.ts:317]: the linter is yelling here. Just because a depency is not updated or do we pass the wrong args here?

**Review by toger5 — APPROVED** [review]

File: src/embedded.ts (hunk 2 above)
Comment (AndrewFerr) [reply src/embedded.ts:317]: It's because of a missing dependency on the updated widget-api.  The linter passes for me locally when `yarn link`ing to the PR-equipped widget-api (other than some errors in an unrelated regtest, which I assume will go away after rebasing onto the latest `develop`...)

**Comment by AndrewFerr** [comment]
I'll take this out of draft once the [dependent widget-api PR](https://github.com/matrix-org/matrix-widget-api/pull/92) is merged.

**Comment by AndrewFerr** [comment]
> I'll take this out of draft once the [dependent widget-api PR](https://github.com/matrix-org/matrix-widget-api/pull/92) is merged.

Not quite -- it would be nice if the widget-api's [release workflow could be fixed](https://github.com/matrix-org/matrix-widget-api/pull/94) before cutting a new release for it, which this PR will depend on.

File: src/embedded.ts (hunk 2 above)
Comment (AndrewFerr) [reply src/embedded.ts:348]: This is indeed fixed now.

**Comment by AndrewFerr** [comment]
After the rebasing, the effective changes since the last review start from 3c65057b.

File: src/embedded.ts [hunk 3]
```diff
@@ -162,6 +182,18 @@ export class RoomWidgetClient extends MatrixClient {
         );
         capabilities.sendToDevice?.forEach((eventType) => widgetApi.requestCapabilityToSendToDevice(eventType));
         capabilities.receiveToDevice?.forEach((eventType) => widgetApi.requestCapabilityToReceiveToDevice(eventType));
+        if (
+            capabilities.sendDelayedEvents &&
+            (capabilities.sendEvent?.length ||
+                capabilities.sendMessage === true ||
+                (Array.isArray(capabilities.sendMessage) && capabilities.sendMessage.length) ||
+                capabilities.sendState?.length)
```
Comment (toger5) [inline src/embedded.ts:190]: Good job on realizing, that requesting this is unnecassary if we dont send any state or message.

**Review by toger5 — APPROVED** [review]
Looks very good

File: src/embedded.ts [hunk 4]
```diff
@@ -280,7 +322,56 @@ export class RoomWidgetClient extends MatrixClient {
...
+    public async _unstable_sendDelayedStateEvent<K extends keyof StateEvents>(
+        roomId: string,
+        delayOpts: SendDelayedEventRequestOpts,
+        eventType: K,
+        content: StateEvents[K],
+        stateKey = "",
+    ): Promise<SendDelayedEventResponse> {
+        if (!(await this.doesServerSupportUnstableFeature(UNSTABLE_MSC4140_DELAYED_EVENTS))) {
+            throw Error("Server does not support the delayed events API");
+        }
```
Comment (robintown) [inline src/embedded.ts:345]: It feels a little surprising that the host client might approve the capability without having server support. (I would expect this check to be the host's responsibility, ideally)

**Review by robintown — APPROVED** [review]
One non-blocking comment:

File: spec/integ/rendezvous/MSC4108SignInWithQR.spec.ts [hunk 5]
```diff
@@ -268,7 +268,8 @@ describe("MSC4108SignInWithQR", () => {
         it("should abort if device doesn't come up by timeout", async () => {
             jest.spyOn(global, "setTimeout").mockImplementation((fn) => {
                 (<Function>fn)();
-                return -1;
+                // TODO: mock timers properly
+                return -1 as any;
```
Comment (t3chguy) [inline spec/integ/rendezvous/MSC4108SignInWithQR.spec.ts:272]: Why are these needed now but weren't before?

Comment (AndrewFerr) [reply spec/integ/rendezvous/MSC4108SignInWithQR.spec.ts:272]: I'm not actually sure...it must be due to a dependency update.  Typechecks expect the return value to be a [NodeJS Timeout](https://nodejs.org/api/timers.html#class-timeout) instead of a timeout ID.

**Review by t3chguy — APPROVED** [review]
